File: hello-world-zr/src/ffi_slice.zig
pub fn FFISlice(comptime T: type) type {
    return extern struct {
        ptr: [*]const T,
        len: usize,
        
        const Self = @This();
        
        pub fn fromSlice(slice: []const T) Self {
            return .{
                .ptr = slice.ptr,
                .len = slice.len,
            };
        }
    };
}

File: hello-world-zr/src/ffi_slice.rs
use std::slice;

#[repr(C)]
pub struct FFISlice<T> {
    ptr: *const T,
    len: usize,
}

impl<T> FFISlice<T> {
    pub fn to_slice(&self) -> &[T] {
        unsafe { slice::from_raw_parts(self.ptr, self.len) }
    }
}
